node/cn/handler_test.go

```diff
@@ -58,3 +62,169 @@ func TestNewProtocolManager(t *testing.T) {
 		assert.Equal(t, errIncompatibleConfig, err)
 	}
 }
+
+func TestSampleSize(t *testing.T) {
+	peers := make([]Peer, minNumPeersToSendBlock-1)
+	assert.Equal(t, len(peers), sampleSize(peers))
+
+	peers = make([]Peer, 4)
+	assert.Equal(t, minNumPeersToSendBlock, sampleSize(peers))
+
+	peers = make([]Peer, 16)
+	assert.Equal(t, 4, sampleSize(peers))
+}
+
+func TestSamplePeersToSendBlock(t *testing.T) {
+	pm := &ProtocolManager{}
+	mockCtrl := gomock.NewController(t)
+
+	block := newBlock(123)
+	hash := block.Hash()
+
+	// 1. When the given node is a consensus node.
+	{
+		// len(pnsWithoutBlock) <= blockReceivingPNLimit
+		mockPeers := NewMockPeerSet(mockCtrl)
+		pm.nodetype = node.CONSENSUSNODE
+		pm.peers = mockPeers
+
+		mockCN := NewMockPeer(mockCtrl)
+		mockPN := NewMockPeer(mockCtrl)
+
+		mockPeers.EXPECT().CNWithoutBlock(hash).Return([]Peer{mockCN}).Times(1)
+		mockPeers.EXPECT().PNWithoutBlock(hash).Return([]Peer{mockPN}).Times(1)
+
+		sampledPeers := pm.samplePeersToSendBlock(block)
+		assert.Equal(t, 2, len(sampledPeers))
+		assert.Equal(t, mockCN, sampledPeers[0])
```

Comment:
Is it important that CN peer enter first when creating peer list in samplePeer sToSendBlock method?  I think it's important that a particular node belongs. What do you think?

Reply:
I didn't consider calling `samplingPeers` inside the `samplePeersToSendBlock`. Therefore, they are returned as-is. Testing `samplingPeers` inside the `samplePeersToSendBlock` seems to be too complicated and out of scope of unit tests. How do you think?